input/full.go

```diff
@@ -20,39 +19,25 @@ import (
 
 // CollectFull - Collects a "full" snapshot of all data we need on a regular interval
 func CollectFull(ctx context.Context, server *state.Server, connection *sql.DB, globalCollectionOpts state.CollectionOpts, logger *util.Logger) (ps state.PersistedState, ts state.TransientState, err error) {
-	systemType := server.Config.SystemType
 	ps.CollectedAt = time.Now()
 
-	ts.Version, err = postgres.GetPostgresVersion(ctx, logger, connection)
+	c, err := postgres.NewCollection(ctx, logger, server, globalCollectionOpts, connection)
 	if err != nil {
-		server.SelfTest.MarkCollectionAspectError(state.CollectionAspectPgVersion, err.Error())
-		logger.PrintError("Error collecting Postgres Version: %s", err)
+		logger.PrintError("Error setting up collection info: %s", err)
 		return
 	}
+	ts.Version = c.PostgresVersion
+	ts.Roles = c.Roles
 
-	if ts.Version.Numeric < state.MinRequiredPostgresVersion {
-		server.SelfTest.MarkCollectionAspectError(state.CollectionAspectPgVersion, "PostgreSQL server version (%s) is too old, 10 or newer is required.", ts.Version.Short)
-		err = fmt.Errorf("Error: Your PostgreSQL server version (%s) is too old, 10 or newer is required.", ts.Version.Short)
-		return
-	}
-
-	server.SelfTest.MarkCollectionAspect(state.CollectionAspectPgVersion, state.CollectionStateOkay, ts.Version.Short)
-
-	ts.Roles, err = postgres.GetRoles(ctx, logger, connection, ts.Version)
-	if err != nil {
-		logger.PrintError("Error collecting pg_roles: %s", err)
-		return
-	}
-
-	ts.Databases, ps.DatabaseStats, err = postgres.GetDatabases(ctx, logger, connection, ts.Version)
+	ts.Databases, ps.DatabaseStats, err = postgres.GetDatabases(ctx, connection)
 	if err != nil {
 		logger.PrintError("Error collecting pg_databases: %s", err)
 		return
 	}
 
 	bufferCacheReady := make(chan state.BufferCache)
 	go func() {
-		postgres.GetBufferCache(ctx, server, globalCollectionOpts, logger, ts.Version, bufferCacheReady)
+		postgres.GetBufferCache(ctx, c, server, globalCollectionOpts, bufferCacheReady)
```

Comment:
Not related to the refactor, but why does this establish its own connection instead of using the one we establish above? :confused:

Reply:
That's intentional since it runs in a separate Go routine (since fetching buffercache stats is slow yet independent of all other data) and we don't have multiplexing enabled on the DB connection.

Reply:
Makes sense. We should probably add a comment explaining that (but not in this PR).